### PR TITLE
Retry if batch results come empty

### PR DESCRIPTION
**Pull request by easafe**
Added an option to `httpGetRequestAndSend` retry downloading batch results if the otherwise correct response body is empty.

### CODE REVIEW

File: lib/force.go [hunk 1]
```diff
@@ -1270,8 +1276,13 @@ func (f *Force) httpGetRequestAndSend(url string, headers map[string]string, res
 				Data:         data,
 			}
 		}
+
 		if err == io.EOF || err == io.ErrUnexpectedEOF {
-			return nil
+			if retryMode == retry && firstChunk && err == io.EOF {
+				f.httpGetRequestAndSend(url, headers, results, doNotRetry)
```
Comment (cwarden) [inline lib/force.go]: The return value is being implicitly discarded.  Should it be this?
```suggestion
				return f.httpGetRequestAndSend(url, headers, results, doNotRetry)
```

Comment (easafe) [reply lib/force.go]: Good catch.

**Comment by cwarden** [comment]
I did more testing, and the error wasn't an empty response, but a malformed chunked response.  The `http` library was returning an "unexpected EOF reading trailer" error.  (It looks like mitmproxy's failure to parse the response made it look like an empty response.)

The changes in f08031cf82b20941f29f0199fbf6aa90800c883a log the error and retry using the method introduced in a03d90b842eb9f6c90035a8ab7b370a278e750d7. But this doesn't fully solve the problem because rather than getting an empty response, we get a partial response, which has already been sent to the channel before the error occurs.  Retrying fetches the entire result set and therefore results in duplicate data being sent to the channel.

The Bulk API doesn't support retrieving `Ranges`.  Therefore, we need to keep track of how many bytes of the result set have been sent to the channel and skip over these bytes if we retry.  I started working on this in 3a328c54791391de5ef800c6ec13831c012d96a5, but I haven't been able to reproduce the error recently, so it may have been a temporary Salesforce issue.

I think we should close this out for now, and see if the error reoccurs.

**Comment by easafe** [comment]
I couldn't reproduce the problem at all- the way I tested it was to manually set a non nil error- maybe you are right in that it was a temporary issue. Retrying seems to be safe only if `firstChunk` is true.

Only tangentially related, but is there any reason why the Force CLI only uses info level for logging? There are a few places that could use warn or error levels, specially if it is being used as a library rather than a standalone program.

**Comment by cwarden** [comment]
On Thu, Jun 27, 2019 at 02:13:39PM -0700, Eduardo Asafe wrote:
> Only tangentially related, but is there any reason why the Force CLI
> only uses info level for logging? There are a few places that could
> use warn or error levels, specially if it is being used as a library
> rather than a standalone program.

Log.Info replaced writing directly to stdout and stderr.

I'm not sure a library should log at the error level.  It should
probably always return an error in these cases.  I'm not sure about
logging warnings from within a library.  Somebody else has probably
given more thought as to how logging within a library should be done.

**Comment by easafe** [comment]
> Log.Info replaced writing directly to stdout and stderr.

Sorry, I mixed up the two terms. It should read "specially if it is being used as a standalone program rather than a library". I wondered it since the object sounds similar to the ones in logging libraries, but looking at log.go I realize it is there so it can used as an interface.
